_posts/2021-04-15-testing-react-tracking-with-jest-and-enzyme.markdown

```diff
@@ -0,0 +1,242 @@
+callback to be passed.
+
+With some help from fellow Artsy engineers Chris Pappas and [Pavlos Vinieratos](https://twitter.com/pvinis), I got
+the tracking to work. Here's how we did it.
+
+<!-- more -->
+
+# A little context
+
+This work took place in Volt, our partner CMS (it's sadly a private repository, but I'll do my best to paste in
+relevant code snippets so you're not totally in the dark). Volt has been around for a long time and has had several
+different client-side tracking implementations over the years. In this case, I wanted to take the opportunity to
+bring Volt up to standard with our other big apps, [Force](https://github.com/artsy/force/) and
+[Eigen](https://github.com/artsy/eigen). They both use react-tracking and
+[Cohesion](https://github.com/artsy/cohesion/), our centralized analytics schema.
+
+Our use-case was a button that would navigate the user to a new page. The button had been implemented in a previous
+PR, and now we wanted to make it execute a tracking call before navigating.
+
+We use Segment for tracking, and their tracking setup relies on a JS snippet being available on your pages. That
+snippet sets a `window.analytics` property, which in turn
+[has a](https://segment.com/docs/connections/sources/catalog/libraries/website/javascript/#track) `.track()`
+method. On a fundamental level, all of our tracking calls boil down to a call to `window.analytics.track()`, where
+we pass a list of properties to `.track()`
+
+# Adding react-tracking
+
+First, there was a bit of setup required to get react-tracking working. The react-tracking package
+[assumes you're using Google Tag Manager by default](https://github.com/NYTimes/react-tracking#custom-optionsdispatch-for-tracking-data),
+but allows you to override that behavior with a custom `dispatch` function. In our case, we wrap our React apps in
+a `<BaseApp>` component, so we added a new `<TrackerContextCohesion>` component with a custom `dispatch` that would
+be available to all of our React apps:
+
+```ts
+interface TrackEventProps {
+  data: { [key: string]: string }
+  options: {}
+  callback: () => void
+}
+
+export const trackEvent = ({ data, options, callback }: Partial<TrackEventProps>): void => {
+  // Action can be something like "click" or "Viewed tooltip"; we pull it out and send it to Segment
+  // separately since it defines the event type
+  const actionName = data.action || data.action_type
+  const trackingData = omit(data, ["action_type", "action"])
+
+  if (actionName) {
+    // This is where we actually call Segment and pass the tracking event
+    window.analytics.track(actionName, trackingData, options, callback)
+  } else {
+    console.error(`Unknown analytics schema being used: ${JSON.stringify(data)}`)
+  }
+}
+
+// We're following the instructions from react-tracking's README on overriding the dispatch function
+export const TrackerContextCohesion = track(
+  {},
```

Comment:
We should fill in this first arg with something just so it doesn't appear mysterious

Reply:
Good thinking! I added a comment  - filling it in felt weird because that wouldn't actually reflect what we're doing here (the first arg would be a tracking call, which we don't want to make when essentially overriding `dispatch`)